### ISSUE TITLE
Support for tasks running on more than 1 architecture

### ISSUE THREAD

**Issue by tommasoboccali** [dmwm/WMCore#10674]
**Impact of the new feature**
This feature is particularly important for running on large HPC systems with architecture != from x86_64, since they will typically be places where no Rucio maintained storage ill be present. Hence, we have the need to be able and run a part of the task (the CPU intensive) on the HPC, and IO intensive parts (merge, cleanup, logcollect) on the site with storage, which will be typically a standard x86_64.

The idea is to submit a task specifying a LIST of architectures it can run on, and then use glidein local matching to accept / refuse parts of the task (for example, the HPC site will veto Merge, Cleanup, logcollect, while the x86 with storage will accept only these).

What is needed and already tested to some extent (see https://indico.cern.ch/event/1049288/) is the ability to
- prepare a classad matching more than one arch
- reconfiguring the payload at runtime from the local environment
- (filter the tipe of subtask -- which is not a WMA feature but a condor one)

**Is your feature request related to a problem? Please describe.**
Not now, but it will be as soon as CMS will have available multiple architecture via (for example) HPC grants

**Describe the solution you'd like**
see https://indico.cern.ch/event/1049288 for a sketched solution.
in principle, the only part missing is the interface by which the user defines the list of possible architectures for a task

**Describe alternatives you've considered**
none (but happy to hear from you!)

**Comment by tommasoboccali** [comment]
a small update: 
we will need to update also the hardwired code in the line

https://github.com/dmwm/WMCore/blob/7900402d03e01d4f3a725fbb48dd102031056318/src/python/WMCore/Storage/Backends/XRDCPImpl.py#L109

and following

**Comment by spigad** [comment]
Hi @amaltaro as per our recent chat I would like to follow-up on this ticket for the implementation of multiarch support.  There are now few use cases  ready to exploit this. The approved grant on M100 first of all ( @tommasoboccali ), then we've Summit which might come soon and ARM is at the horizon.  Would you consider raise the priority on this ?

Of course we are available to test anything that you think could help you here. Just let us know. 

Thanks Daniele

**Comment by spigad** [comment]
Hi @amaltaro is this gonna be considered  in te list of possible candidates for the next quarter?  Is there anything more I should do to help n this process ? 
thanks a lot for your support. 
cheers Daniele

**Comment by amaltaro** [comment]
@spigad Hi Daniele, thanks for the reminder.

We will have to refactor this method (such that it can return a list of architectures):
https://github.com/dmwm/WMCore/blob/540994ab2f751c7f584522c8af02784756ccfe5e/src/python/WMCore/BossAir/Plugins/BasePlugin.py#L156

and adapt the `Requirements` condor expression to support a list/multiple architectures to be requested, around these lines:
https://github.com/dmwm/WMCore/blob/master/src/python/WMCore/BossAir/Plugins/SimpleCondorPlugin.py#L631-L635

@mmascher @aperezca do you know if multiple architectures are supported in glideinWMS and would you have any recommendation on how to construct that expression (see SimpleCondorPlugin above)?

**Comment by mmascher** [comment]
Hi @amaltaro , right now there are no entries in the factories other than x86 ones. Marconi100 are manually launched glideins, so they do not need a CE. Therefore we never discussed how to deal with the "first matchmaking" for non x86 job/resources. Luckily this is something as simple as modifying the match expression in the frontend (no development needed).

My proposal would be to add something like `job.get("REQUIRED_ARCH") == glidein["attrs"]["GLIDEIN_REQUIRED_ARCH"` in the frontend match expression (or possibly `glidein["attrs"]["GLIDEIN_REQUIRED_ARCH" in job.get("REQUIRED_ARCH")` if list of archs are required).

Now, I think I digressed as the question was about Marconi100 and the **second** matchmaking. According to the condor documentation (https://htcondor.readthedocs.io/en/latest/classad-attributes/job-classad-attributes.html):

```If the job requirements does not mention the (startd) attribute OPSYS, the schedd will append a clause to Requirements forcing the job to match the same OPSYS as the submit machine. The schedd appends a simliar clause to match the ARCH```

In other words you can do whatever you want with the Requirement expression here https://github.com/dmwm/WMCore/blob/master/src/python/WMCore/BossAir/Plugins/SimpleCondorPlugin.py#L631-L635
As long as you mention TARGET.Arch condor will not add any other requirement anymore (i.e.: `stringListMember(TARGET.Arch, %s)`). Does it make sense?

**Comment by khurtado** [comment]
@mmascher : 

I followed your suggestion adding `stringListMember(TARGET.Arch, %s)`,  

https://github.com/dmwm/WMCore/blob/d06f34d39724ebc0fe77e18a2ad9c38afa076495/src/python/WMCore/BossAir/Plugins/SimpleCondorPlugin.py#L635

but it seems condor. still added its own `TARGET.Arch` requirements after that. 

```
        Requirements = (stringListMember(TARGET.Arch, "X86_64,ppc64le") && (TARGET.Arch == "X86_64") && (TARGET.OpSys == "LINUX") && (TARGET.Disk >= RequestDisk) && (TARGET.Memory >= RequestMemory) && (TARGET.Cpus >= RequestCpus) && (TARGET.HasFileTransfer)
```

Any suggestions?

**Comment by spigad** [comment]
Hi @khurtado I gave it a try with a single arch as I was doing just before your patch on vocms0263 and I see that submission if failing [*] for the same reason you highliglited above.  I fear something has been broken with your patch.

This was all working nicely before .. a example is here: spiga_TC_SLC7_Marconi_TB_CMS_Marconi_220728_094249_8414
this was the last one working.

```
ComponentLog:	Requirements = (stringListMember(TARGET.Arch, "ppc64le") && (TARGET.Arch == "X86_64") && (TARGET.OpSys == "LINUX") && (TARGET.Disk >= RequestDisk) && (TARGET.Memory >= RequestMemory) && (TARGET.Cpus >= RequestCpus) && (TARGET.HasFileTransfer)
```

**Comment by khurtado** [comment]
@spigad Yes, at first I thought it was a condor error, but it seems it might just have been a typo. I have restarted the agent with a small fix.

**Comment by spigad** [comment]
@khurtado I've injected a new WF (```spiga_TC_SLC7_Marconi_TB_CMS_Marconi_220731_160802_6847``` ) and  it works now [*]
 however I would have expected to see some pressure also at Tier1-CNAF where those jobs could run. I see notihing instead. This is something to verify with SI people. 

[*]
```
Requirements = (stringListMember(TARGET.Arch,"X86_64,ppc64le")) && (TARGET.OpSys == "LINUX") && (TARGET.Disk >= RequestDisk) && (TARGET.Memory >= RequestMemory) && (TARGET.Cpus >= RequestCpus) && (TARGET.HasFileTransfer)
```

**Comment by khurtado** [comment]
@spigad But I do see jobs running on different entries, including Marconi100 in the job history:

``` 
$ condor_history -constraint 'regexp("spiga_TC_SLC7_Marconi_TB_CMS_Marconi_220731_160802_6847", Args)' -limit 50 -af MATCH_EXP_JOB_Site ExitCode | sort | uniq -c
     27 CMSHTPC_T1_IT_CNAF_CINECA_Marconi100_gpu 0
      9 CMSHTPC_T1_IT_CNAF_CINECA_Marconi100_gpu 85
      3 CNAF 0
      4 CNAF 152
      7 CNAF 81
```

**Comment by spigad** [comment]
you're right was about to comment bach heere!! I've checked the condor logs and I see the same as you ! thanks

it seems we're  colse to the target. Thanks a lot
Daniele

**Comment by mmascher** [comment]
Back to this after the vacations, are you also adding a `REQUIRED_ARCH` classad in WMAgent that we could use for the first matchmaking in glideinWMS? To request pilots and provision the resources.

**Comment by amaltaro** [comment]
@mmascher Hi Marco, sorry for the very late reply. We can start discussing this in a new ticket: https://github.com/dmwm/WMCore/issues/11291
